src/hooks/preparse.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import type { Hook } from '@oclif/core';
+
+const hook: Hook<'preparse'> = async function ({ argv, options, context }) {
+  // Skip this hook if command does not have a --flags-dir flag or if it is not present in argv
+  if (!argv.includes('--flags-dir') || !options.flags?.['flags-dir']) return argv;
+
+  const { default: fs } = await import('node:fs/promises');
+  const { default: path } = await import('node:path');
+  context.debug('Initial argv', argv.join(' '));
+  const flagsToIgnore = new Set(
+    Object.entries(options.flags ?? {})
+      .filter(
+        ([, flagOptions]) =>
+          // don't ignore if flag can take multiple values
+          (flagOptions.type === 'option' && flagOptions.multiple !== true) || flagOptions.type === 'boolean'
+      )
+      .filter(
+        ([flagName, flagOptions]) =>
+          // ignore if short char flag is present
+          argv.includes(`-${flagOptions.char}`) ||
+          // ignore if long flag is present
+          argv.includes(`--${flagName}`) ||
+          // ignore if --no- flag is present
+          (flagOptions.type === 'boolean' && flagOptions.allowNo && argv.includes(`--no-${flagName}`))
+      )
+      .flatMap(([flagName, flagOptions]) => {
+        // Also ignore the --no- flag if boolean flag allows it
+        if (flagOptions.type === 'boolean' && flagOptions.allowNo) {
+          return [flagName, `no-${flagName}`];
+        }
+
+        return [flagName];
+      })
+  );
+  context.debug('Flags to ignore', flagsToIgnore);
+
+  const flagsDir = argv[argv.indexOf('--flags-dir') + 1];
+  if (!flagsDir) {
+    context.debug('No flags dir provided');
+    return argv;
+  }
+  if (flagsDir.startsWith('-')) {
+    context.debug(`No flags dir provided, got ${flagsDir}`);
+    return argv;
+  }
```

Comment:
these 2 checks could go right after line 11 (so that we do all the short-circuit checks before the more expensive imports and flagsToIgnore calculcation)